packages/expo-haptics/ios/HapticsModule.swift

```diff
@@ -66,3 +82,129 @@ public class HapticsModule: Module {
+    case "hapticSharpness": self = .hapticSharpness
+    case "attackTime": self = .attackTime
+    case "decayTime": self = .decayTime
+    case "releaseTime": self = .releaseTime
+    case "sustained": self = .sustained
+    default:
+      throw NSError(domain: "ExpoHaptics", code: 2, userInfo: [NSLocalizedDescriptionKey: "Unknown parameter ID: \(string)"])
+    }
+  }
+}
+
+extension CHHapticDynamicParameter.ID {
+  init(from string: String) throws {
+    switch string {
+    case "hapticIntensityControl": self = .hapticIntensityControl
+    case "hapticSharpnessControl": self = .hapticSharpnessControl
+    default:
+      throw NSError(domain: "ExpoHaptics", code: 3, userInfo: [NSLocalizedDescriptionKey: "Unknown dynamic parameter ID: \(string)"])
+    }
+  }
+}
+
+extension HapticEventParameter {
+  func toCHHapticParameter() throws -> CHHapticEventParameter {
+    let id = try CHHapticEvent.ParameterID(from: parameterID)
+    return CHHapticEventParameter(parameterID: id, value: value)
+  }
+}
+
+extension HapticEvent {
+  func toCHHapticEvent() throws -> CHHapticEvent {
+    let type = try CHHapticEvent.EventType(from: eventType)
+    let params = try (parameters ?? []).map { try $0.toCHHapticParameter() }
+    let timeValue = time ?? 0.0
+    if let duration = eventDuration {
+      return CHHapticEvent(eventType: type, parameters: params, relativeTime: timeValue, duration: duration)
+    } else {
+      return CHHapticEvent(eventType: type, parameters: params, relativeTime: timeValue)
+    }
+  }
+}
+
+extension HapticParameterCurveControlPoint {
+  func toCHHapticParameterCurveControlPoint() -> CHHapticParameterCurve.ControlPoint {
+    CHHapticParameterCurve.ControlPoint(relativeTime: relativeTime, value: value)
+  }
+}
+
+extension HapticParameterCurve {
+  func toCHHapticParameterCurve() throws -> CHHapticParameterCurve {
+    let id = try CHHapticDynamicParameter.ID(from: parameterID)
+    let points = (controlPoints ?? []).map { $0.toCHHapticParameterCurveControlPoint() }
+    let startTime = relativeTime ?? 0.0
+    return try CHHapticParameterCurve(parameterID: id, controlPoints: points, relativeTime: startTime)
+  }
+}
+
+extension HapticPatternData {
+  func toCHHapticPattern() throws -> CHHapticPattern {
+    let events = try self.events.map { try $0.toCHHapticEvent() }
+    let curves = try (self.parameterCurves ?? []).map { try $0.toCHHapticParameterCurve() }
+    return try CHHapticPattern(events: events, parameterCurves: curves)
+  }
+}
+
+private extension HapticsModule {
+  func initializeHapticEngine() {
+    guard CHHapticEngine.capabilitiesForHardware().supportsHaptics else { return }
+    do {
+      hapticEngine = try CHHapticEngine()
+      try hapticEngine?.start()
+    } catch {
+      print("ExpoHaptics: Failed to initialize haptic engine: \(error)")
+    }
+  }
+
+  func cleanupHapticEngine() {
+    hapticEngine?.stop(completionHandler: nil)
+    hapticEngine = nil
+  }
+}
```

Comment:
Could you move this to `HapticsModule`? We generally try to avoid using extensions for classes/structs within the same package. They are useful for:
- extending external types, i.e. platform and 3rd-party types
- providing default implementation for protocols

The same applies to the extensions for records.